python/neuroglancer/ingest/tasks.py

```diff
@@ -1,66 +1,348 @@
+        return "IngestTask(chunk_path='{}', chunk_encoding='{}', info_path='{}'')".format(
+            self.chunk_path, self.chunk_encoding, self.info_path)
+
+    def execute(self):
+        self._parse_chunk_path()
+        self._parse_info_path()
+        self._storage = Storage(self._dataset_name, self._layer_name, compress=True)
+        self._download_info()
+        self._download_data()
+        self._create_chunks()
+
+    def _parse_chunk_path(self):
+        match = re.match(r'^.*/(\d+)-(\d+)_(\d+)-(\d+)_(\d+)-(\d+)$', self.chunk_path)
+        (self._xmin, self._xmax,
+         self._ymin, self._ymax,
+         self._zmin, self._zmax) = map(int, match.groups())
+        self._filename = "{}-{}_{}-{}_{}-{}".format(
+            self._xmin, self._xmax,
+            self._ymin, self._ymax,
+            self._zmin, self._zmax)
+
+    def _parse_info_path(self):
+        match = re.match(r'^.*/([^//]+)/([^//]+)/info$', self.info_path)
+        self._dataset_name, self._layer_name = match.groups()
+
+    def _download_info(self):
+        info_string = self._storage.get_blob(
+                '{}/{}/info'.format(self._dataset_name, self._layer_name) \
+            ).download_as_string()
+        self._info = json.loads(info_string)
+
+    def _download_data(self):
+        string_data = self._storage.get_blob(
+            '{}/{}/build/{}'.format(self._dataset_name, self._layer_name, self._filename)) \
+        .download_as_string()
+
+        if self.chunk_encoding == 'npz':
+          self._data = chunks.decode_npz(string_data)
+        else:
+          raise NotImplementedError(self.chunk_encoding)
 
-    @chunk_path.setter
-    def chunk_path(self, value):
-        self._chunk_path = value
+    def _create_chunks(self):
+        for scale in self._info["scales"]:
+            for chunk_size in scale['chunk_sizes']:
+                for encoded, filename in self._generate_chunks(scale, chunk_size):
+                    self._storage.add_file(filename, encoded)
+            self._storage.flush(scale['key'])
+
+    def _generate_chunks(self, scale, chunk_size):
+        highest_resolution = np.array(self._info['scales'][0]['resolution'])
+        current_resolution = np.array(scale["resolution"])
+        downsample_ratio = current_resolution / highest_resolution
+        current_shape = self._data.shape[:-1] / downsample_ratio # we discard the channel component in data
+
+        # This is required because last data chunk is allowed to not 
+        # be a multiple of the chunk size.
+        n_chunks = self._ceil(current_shape, chunk_size)
+        for x,y,z in itertools.product(*list(map(xrange, n_chunks))):
+            # numpy allows for index that are larger than the array size
+            # because we are doing ceil in n_chunks, there will be cases where
+            # chunk.shape is less than chunk_size. In that case the filename should 
+            # base on chunk.shape.
+            chunk = self._data[
```

Comment:
might want to consider using neuroglancer.chunks.downsample.downsample_with_striding, that way we avoid copy pasting this downsample code again and again